src/parquet/util/buffer.cc

```diff
@@ -34,18 +35,68 @@ std::shared_ptr<Buffer> MutableBuffer::GetImmutableView() {
   return std::make_shared<Buffer>(this->get_shared_ptr(), 0, size());
 }
 
-OwnedMutableBuffer::OwnedMutableBuffer() :
-    ResizableBuffer(nullptr, 0) {}
+template <class T>
+OwnedMutableBuffer<T>::OwnedMutableBuffer(int64_t size, MemoryPool* pool) :
+    ResizableBuffer(nullptr, 0), memory_pool_(pool) {
+  Resize(size);
+}
+
+template <class T>
+OwnedMutableBuffer<T>::~OwnedMutableBuffer() {
+  if (mutable_data_) {
+    memory_pool_->Free(mutable_data_);
+  }
+}
+
+template <class T>
+void OwnedMutableBuffer<T>::Reserve(int64_t new_capacity) {
+  if (!mutable_data_ || new_capacity > capacity_) {
+    if (mutable_data_) {
+      uint8_t* new_data = memory_pool_->Allocate(new_capacity * sizeof(T));
+      memcpy(new_data, mutable_data_, size_ * sizeof(T));
+      memory_pool_->Free(mutable_data_);
+      mutable_data_ = new_data;
+    } else {
+      mutable_data_ = memory_pool_->Allocate(new_capacity * sizeof(T));
+    }
+    data_ = mutable_data_;
+    buffer_ = reinterpret_cast<T*>(mutable_data_);
+    capacity_ = new_capacity;
+  }
+}
 
-void OwnedMutableBuffer::Resize(int64_t new_size) {
+template <class T>
+void OwnedMutableBuffer<T>::Resize(int64_t new_size) {
+  Reserve(new_size);
   size_ = new_size;
-  try {
-    buffer_owner_.resize(new_size);
-  } catch (const std::bad_alloc& e) {
-    throw ParquetException("OOM: resize failed");
+}
+
+template <class T>
+void OwnedMutableBuffer<T>::Clear() {
+  Resize(0);
+}
+
+template <class T>
+T& OwnedMutableBuffer<T>::operator[](int64_t i) {
+  return buffer_[i];
+}
+
+template <class T>
+void OwnedMutableBuffer<T>::Assign(int64_t size, const T val) {
+  Resize(size);
+  for (int64_t i = 0; i < size_; i++) {
+    buffer_[i] = val;
   }
-  data_ = buffer_owner_.data();
-  mutable_data_ = buffer_owner_.data();
 }
 
+template class OwnedMutableBuffer<uint8_t>;
+template class OwnedMutableBuffer<int32_t>;
+template class OwnedMutableBuffer<int64_t>;
+template class OwnedMutableBuffer<bool>;
+template class OwnedMutableBuffer<float>;
+template class OwnedMutableBuffer<double>;
+template class OwnedMutableBuffer<Int96>;
+template class OwnedMutableBuffer<ByteArray>;
+template class OwnedMutableBuffer<FixedLenByteArray>;
```

Comment:
At the moment, the only place where `T` is not `uint8_t` is in the dict encoder hash table implementation. I'm not arguing against having an `std::vector` replacement, only that it would be better to do it using object composition (so no inheritance) or implementing a custom STL allocator. This doesn't strike me as code duplication since it is introducing a new functional concept (typed value access and other APIs) to the API.